### DISCUSSION TITLE
Scoped API tokens

### DISCUSSION THREAD

**Discussion by mlissner** [plausible/analytics#1767] [Anything Else]
I was a bit surprised when setting up an API key that it has global permission across all sites and endpoints. 

We have one site where the dashboard and stats are public and another where they're private. It'd be nice to be able to share the API key for the site with public stats without worrying about it also having access to the private stats, but the only way I can think to do that would be by creating multiple accounts (which is a pretty big hack!).

For public sites, it also seems like the API key is pretty harmless, but I'm not sure and I'd worry that eventually an API key would become read/write, even if it's a purely read-only API now (is it? I'm not sure). Anyhow, it'd be nice to scope a key per endpoint, so that we could safely share a key for the breakdown API on a particular site, say.

This is just a small detail, but something to think about if/when something like this ever comes along might be to add the site name to the key itself. For example, instead of a key like:

```
asdflakdjfkalsdfkjasldfajsdfklasjfkalsjfkslajf
```

You could have:

```
google.com-asdfalsdkfjasdflajskdfjaksldfkjaskdfj
```

Anyhow, just some ideas! Thanks for the great tool.

**Comment by ukutaht** [comment]
Thanks @mlissner. I anticipated this point but I cut the scope for this to get the API out quicker. We actually store a `scope` field but at the moment it's always something like `stats:read:*` or `sites:provision:*`. The idea was that in the future it could be like `stats:read:sitea.com,siteb.com` or whatever. This isn't implemented but it's like 70% there on the backend.

As you can see the stats api is meant to be readonly and it will stay that way forever. We have a separate Events API and Sites API that have write permissions.

So yeah, sorry for not having it yet. It will get implement at some point but I can't tell when.

**Reply by mlissner** [reply]
Love it, that's great, thanks.
